### PR TITLE
Implement XML encryption

### PR DESCRIPTION
**Pull request by jaimeperez**
This PR follows previous developments to implement XML encryption in an analogous way to existing digital signature code.

### CODE REVIEW

File: src/Alg/AbstractAlgorithmFactory.php [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+     * is blacklisted, unknown or the given key is not suitable for it.
+     */
+    public function getAlgorithm(string $algId, AbstractKey $key): AlgorithmInterface
+    {
+        Assert::true(
+            !in_array($algId, $this->blacklist, true),
+            sprintf('Blacklisted algorithm: \'%s\'.', $algId),
+            BlacklistedAlgorithmException::class
+        );
+        Assert::true(
```
Comment (tvdijen) [inline src/Alg/AbstractAlgorithmFactory.php]: Assert::keyExists(static::$cache, $algId, [..]);

File: src/Alg/AbstractAlgorithmFactory.php [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+     * @param \SimpleSAML\XMLSecurity\Key\AbstractKey $key The key to use with the given algorithm.
+     *
+     * @return \SimpleSAML\XMLSecurity\Alg\AlgorithmInterface An object implementing the given algorithm.
+     *
+     * @throws \SimpleSAML\XMLSecurity\Exception\InvalidArgumentException If an error occurs, e.g. the given algorithm
+     * is blacklisted, unknown or the given key is not suitable for it.
+     */
+    public function getAlgorithm(string $algId, AbstractKey $key): AlgorithmInterface
+    {
+        Assert::true(
```
Comment (tvdijen) [inline src/Alg/AbstractAlgorithmFactory.php:78]: Assert::notInArray($algId, $this->blacklist, [..]);

File: src/Alg/Encryption/AES.php [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    {
+        parent::__construct($key, $algId);
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public static function getSupportedAlgorithms(): array
+    {
+        return [
```
Comment (tvdijen) [inline src/Alg/Encryption/AES.php:38]: Was 3DES left out intentionally, can we use `array_keys(C::BLOCK_CIPHER_ALGORITHMS)` instead?

File: src/Alg/Encryption/AbstractEncryptor.php [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+     * @return string
+     */
+    public function getAlgorithmId(): string
+    {
+        return $this->algId;
+    }
+
+
+    /**
+     * @return AbstractKey
```
Comment (tvdijen) [inline src/Alg/Encryption/AbstractEncryptor.php]: Use full namespace

File: src/Alg/Signature/SignatureAlgorithmFactory.php [hunk 5]
```diff
@@ -4,32 +4,30 @@
...
-     * An array holding the known classes extending \SimpleSAML\XMLSecurity\Alg\Signature\AbstractSigner.
+     * A cache of algorithm implementations indexed by algorithm ID.
      *
      * @var string[]
      */
-    private static array $algorithms = [
-        RSA::class,
-        HMAC::class
-    ];
+    protected static array $cache = [];
```
Comment (tvdijen) [inline src/Alg/Signature/SignatureAlgorithmFactory.php:23]: Why the change from private to protected? The class is final anyway.

File: src/Alg/Signature/SignatureAlgorithmFactory.php [hunk 6]
```diff
@@ -38,44 +36,37 @@ final class SignatureAlgorithmFactory
      *
      * @var string[]
      */
-    private array $blacklist = [
+    protected array $blacklist = [
```
Comment (tvdijen) [inline src/Alg/Signature/SignatureAlgorithmFactory.php:39]: same

File: src/Alg/Encryption/AES.php (hunk 3 above)
Comment (jaimeperez) [reply src/Alg/Encryption/AES.php:38]: 3DES is... 3DES, not AES 😛

File: src/Backend/OpenSSL.php [hunk 7]
```diff
@@ -51,6 +51,12 @@ final class OpenSSL implements EncryptionBackend, SignatureBackend
     /** @var int */
     protected int $keysize;
 
+    /** @var bool */
+    protected bool $useAuthTag = false;
+
+    /** @var int */
+    const AUTH_TAG_LEN = 16;
```
Comment (tvdijen) [inline src/Backend/OpenSSL.php]: declare visibility for const.

File: src/Alg/AbstractAlgorithmFactory.php (hunk 1 above)
Comment (jaimeperez) [reply src/Alg/AbstractAlgorithmFactory.php]: Hey! This one is on you! (I just basically took what was in a previous commit in master 😛). But the good thing is that now with proper autocompletion in my IDE for the `Assert` lib, this is easy-peasy.

File: src/Alg/Encryption/AES.php (hunk 3 above)
Comment (tvdijen) [reply src/Alg/Encryption/AES.php:38]: I know, but it doesn't mean it's not supported.. You're basically by-passing the blacklist if you leave it out here.

File: src/Alg/AbstractAlgorithmFactory.php (hunk 1 above)
Comment (jaimeperez) [reply src/Alg/AbstractAlgorithmFactory.php]: Actually, do we have the  `not*` variants? Can I do `Assert::notKeyExists()`? Because the condition is negated here...

Comment (jaimeperez) [reply src/Alg/AbstractAlgorithmFactory.php]: Nevermind 😅

File: src/XML/xenc/AbstractEncryptedType.php [hunk 8]
```diff
@@ -56,7 +56,7 @@ class AbstractEncryptedType extends AbstractXencElement
      *   The EncryptionMethod object of this EncryptedData. Optional.
      * @param \SimpleSAML\XMLSecurity\XML\ds\KeyInfo|null $keyInfo The KeyInfo object of this EncryptedData. Optional.
      */
-    protected function __construct(
+    public function __construct(
```
Comment (tvdijen) [inline src/XML/xenc/AbstractEncryptedType.php:59]: Should it be abstract?

Comment (tvdijen) [reply src/XML/xenc/AbstractEncryptedType.php:59]: Also, the phpdoc says `EncryptedData constructor`, which is incorrect

File: src/XML/xenc/EncryptedKey.php [hunk 9]
```diff
@@ -125,6 +125,84 @@ protected function setReferenceList(?ReferenceList $referenceList): void
...
+     * to decrypt the key.
+     *
+     * @return string The decrypted key.
+     */
+    public function decrypt(EncryptionAlgorithmInterface $decryptor): string
+    {
+        if ($this->getCipherData()->getCipherValue() === null) {
+            throw new InvalidArgumentException('Decrypting keys by reference is not supported.');
+        }
+
```
Comment (tvdijen) [inline src/XML/xenc/EncryptedKey.php:141]: Maybe replace both if-statements with Asserts?

File: src/XML/xenc/EncryptedKey.php [hunk 10]
```diff
@@ -125,6 +125,84 @@ protected function setReferenceList(?ReferenceList $referenceList): void
...
+        ?string $id = null,
+        ?string $type = null,
+        ?string $mimeType = null,
+        ?string $encoding = null,
+        ?string $recipient = null,
+        ?CarriedKeyName $carriedKeyName = null,
+        ?KeyInfo $keyInfo = null,
+        ?ReferenceList $referenceList = null
+    ): EncryptedKey {
+        if ($encryptor->getAlgorithmId() !== $encryptionMethod->getAlgorithm()) {
```
Comment (tvdijen) [inline src/XML/xenc/EncryptedKey.php]: Use Assert?

**Review by tvdijen — APPROVED** [review]

File: src/Alg/AbstractAlgorithmFactory.php (hunk 2 above)
Comment (jaimeperez) [reply src/Alg/AbstractAlgorithmFactory.php:78]: This one can't do, since `Assert::notInArray()` does not exist.

File: src/Alg/Encryption/AbstractEncryptor.php (hunk 4 above)
Comment (jaimeperez) [reply src/Alg/Encryption/AbstractEncryptor.php]: Done!

File: src/Backend/OpenSSL.php (hunk 7 above)
Comment (jaimeperez) [reply src/Backend/OpenSSL.php]: Done!

File: src/Alg/Signature/SignatureAlgorithmFactory.php (hunk 5 above)
Comment (jaimeperez) [reply src/Alg/Signature/SignatureAlgorithmFactory.php:23]: It needs to be protected to be accessible from the abstract parent class.

File: src/Alg/Signature/SignatureAlgorithmFactory.php (hunk 6 above)
Comment (jaimeperez) [reply src/Alg/Signature/SignatureAlgorithmFactory.php:39]: Same reason. If private, the abstract class won't be able to access it.

File: src/XML/xenc/AbstractEncryptedType.php (hunk 8 above)
Comment (jaimeperez) [reply src/XML/xenc/AbstractEncryptedType.php:59]: Why? We have an implementation here... Do you mean maybe the class itself? Because that one is not declared abstract despite its name, and it should be. Changing it 😉

File: src/XML/xenc/EncryptedKey.php (hunk 9 above)
Comment (jaimeperez) [reply src/XML/xenc/EncryptedKey.php:141]: Consider it done!

File: src/XML/xenc/EncryptedKey.php (hunk 10 above)
Comment (jaimeperez) [reply src/XML/xenc/EncryptedKey.php]: Done too!

**Comment by tvdijen** [comment]
You may want to finish the README as part of this PR as well....

File: README.md [hunk 11]
```diff
@@ -286,7 +286,307 @@ forces you to **remember that you need to check the key used to verify the signa
...
 ## Encryption API
 
-Not available yet.
+The XML encryption API is similar to its signature counterpart, and also consists of two main
+interfaces:
+
+- `SimpleSAML\XMLSecurity\XML\EncryptableElementInterface`
+- `SimpleSAML\XMLSecurity\XML\EncryptedElementInterface`
+
+Just like in the signature APi, the former signals that an object can be encrypted (and as such
```
Comment (tvdijen) [inline README.md]: Uppercase `i`

Comment (jaimeperez) [reply README.md]: Done!
